src/babel/transformation/modules/system.js

```diff
@@ -163,6 +163,48 @@ export default class SystemFormatter extends AMDFormatter {
 
       return t.functionExpression(null, [uid], t.blockStatement(state.nodes));
     }));
+
+    this._optimizeSetters(scope, setters);
+
+    return setters;
+  }
+
+  /**
+   * Setters are optimized to avoid slow export behavior in modules that rely on deep hierarchies
+   * of export-from declarations.
+   * More info in https://github.com/babel/babel/pull/1722 and
+   * https://github.com/ModuleLoader/es6-module-loader/issues/386.
+   *
+   * TODO: Ideally this would be optimized during construction of the setters, but the current
+   * architecture of the module formatters make that difficult.
+   */
+  _optimizeSetters(scope, setters){
+    var hasExports;
+    var exportObjectIdentifier;
+    scope.traverse(setters, {
```

Comment:
The visitor should be hoisted to the top scope and `state` should be used instead. Otherwise normalisation of the visitor will happen each time and there's a lot of wasted allocation.